Sources/NetworkProtection/PacketTunnelProvider.swift

```diff
@@ -756,65 +762,60 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
         }
 
         do {
-            let tunnelConfiguration = try await generateTunnelConfiguration(environment: environment,
-                                                                            serverSelectionMethod: serverSelectionMethod,
+            let tunnelConfiguration = try await generateTunnelConfiguration(serverSelectionMethod: serverSelectionMethod,
                                                                             includedRoutes: includedRoutes ?? [],
                                                                             excludedRoutes: settings.excludedRanges,
                                                                             regenerateKey: regenerateKey)
 
-            try await withCheckedThrowingContinuation { [weak self] (continuation: CheckedContinuation<Void, Error>) in
-                guard let self = self else {
-                    continuation.resume()
+            try await updateAdapterConfig(tunnelConfiguration: tunnelConfiguration, reassert: reassert)
+
+            providerEvents.fire(.tunnelUpdateAttempt(.success))
+        } catch {
+            providerEvents.fire(.tunnelUpdateAttempt(.failure(error)))
+        }
+    }
+
+    @MainActor
+    private func updateAdapterConfig(tunnelConfiguration: TunnelConfiguration, reassert: Bool) async throws {
```

Comment:
This diff is a bit yucky, but I just pulled out the updating of the adapter config into a helper function as it’s needed for the failure recovery.

Reply:
A few notes about this change:
- Adding a new, more generic method is fine.
- You should move the tunnel update pixels to be fired here, so the failure recovery will still fire tunnel update pixels.
- I'd consider naming that's more consistent with existing methods: `updateAdapterConfig` > `updateAdapterConfiguration`.